heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```diff
@@ -0,0 +1,298 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.ckptmgr;
+
+import java.io.IOException;
+import java.nio.channels.SocketChannel;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Message;
+
+import com.twitter.heron.common.basics.NIOLooper;
+import com.twitter.heron.common.network.HeronServer;
+import com.twitter.heron.common.network.HeronSocketOptions;
+import com.twitter.heron.common.network.REQID;
+import com.twitter.heron.proto.ckptmgr.CheckpointManager;
+import com.twitter.heron.proto.system.Common;
+import com.twitter.heron.spi.statefulstorage.Checkpoint;
+import com.twitter.heron.spi.statefulstorage.IStatefulStorage;
+import com.twitter.heron.spi.statefulstorage.StatefulStorageException;
+
+public class CheckpointManagerServer extends HeronServer {
+  private static final Logger LOG = Logger.getLogger(CheckpointManagerServer.class.getName());
+
+  private final String topologyName;
+  private final String topologyId;
+  private final String checkpointMgrId;
+  private final IStatefulStorage checkpointsBackend;
+
+  private SocketChannel connection;
+
+  public CheckpointManagerServer(
+      String topologyName, String topologyId, String checkpointMgrId,
+      IStatefulStorage checkpointsBackend, NIOLooper looper, String host,
+      int port, HeronSocketOptions options) {
+    super(looper, host, port, options);
+
+    this.topologyName = topologyName;
+    this.topologyId = topologyId;
+    this.checkpointMgrId = checkpointMgrId;
+    this.checkpointsBackend = checkpointsBackend;
+
+    this.connection = null;
+
+    registerInitialization();
+  }
+
+  private void registerInitialization() {
+    registerOnRequest(CheckpointManager.RegisterStMgrRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.RegisterTMasterRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.SaveInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.GetInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.CleanStatefulCheckpointRequest.newBuilder());
+  }
+
+  @Override
+  public void onConnect(SocketChannel channel) {
+    LOG.info("Got a new connection from host:port "
+        + channel.socket().getRemoteSocketAddress());
+  }
+
+  @Override
+  public void onRequest(REQID rid, SocketChannel channel, Message request) {
+    if (request instanceof CheckpointManager.RegisterStMgrRequest) {
```

Comment:
Thrift allows the ability to do something like this (psuedo-code), which is really convenient for thinks like this. It doesn't seem like protobufs allows this though. It seems to limit the enum values to just constants, which is unfortunate.
```
enum RequestTypes {
  RegisterStMgrRequest,
  RegisterTMasterRequest
  ...
}
```